### ISSUE TITLE
Preprocessing bottleneck in YOLOv8 Classification

### ISSUE THREAD

**Issue by Sachin-Wani** [ultralytics/ultralytics#13435]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

1.	**Is there a way to speed up pre-processing bottleneck?** The model is running on CUDA. However (in my opinion), the pre-processing of taking the video, getting its frame, converting it from 2688x1520 to 640x640 and setting it up as input to the model takes place in the CPU and becomes the bottleneck.
2.	**Should I retrain the model with better and/or different arguments?** I thought rectangular sized images/frames is slower to inference hence I went with default 640x640 training and inferencing. Should I try 16x smaller resolution: 672x380 (instead of 2688x1520) for rectangular training and inference.

### Questions again at the bottom. Providing info here.

Hello! I am training a YOLOv8s-cls classifier model for binary classification for video classification (frame-by-frame). The videos are 2688x1520 @30fps
To create the dataset, I split the videos into individual frame images and put them into two folders to train my model. I used about 20000 images per class for train and 2000 for val. 

### I trained my model using the following parameters:

```
MODEL_TYPE = 'yolov8s-cls.pt'
model = YOLO(MODEL_TYPE)
model.train(data=DATA_DIR, epochs=50, imgsz = 640, batch = 32,
            single_cls = True, dropout = 0.1, dfl = 8, 
            degrees = 20, translate = 0.15, mosaic = 0, 
            device = 0)
```

#### At the end of training epochs:

>       Epoch    GPU_mem       loss  Instances       Size
>       50/50       4.5G    0.01264          7        640: 100%|██████████| 1233/1233 [07:10<00:00,  2.87it/s]
>                classes   top1_acc   top5_acc: 100%|██████████| 30/30 [00:12<00:00,  2.49it/s]
>                    all      0.971          1
> 
> 
> 50 epochs completed in 6.098 hours.

#### Validating model (which is just images/frames):

> ultralytics.utils.metrics.ClassifyMetrics object with attributes:
> 
> confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7f6bec1f4430>
> curves: []
> curves_results: []
> fitness: 0.9886122941970825
> keys: ['metrics/accuracy_top1', 'metrics/accuracy_top5']
> results_dict: {'metrics/accuracy_top1': 0.977224588394165, 'metrics/accuracy_top5': 1.0, 'fitness': 0.9886122941970825}
> save_dir: PosixPath('runs/classify/train4')
> speed: {'preprocess': 0.4146106667437796, 'inference': 0.34963238542362796, 'loss': 0.00014484433804528188, 'postprocess': 0.00013701491436715854}
> task: 'classify'
> top1: 0.977224588394165
> top5: 1.0
> Speed: 0.4ms preprocess, 0.6ms inference, 0.0ms loss, 0.0ms postprocess per image

#### When running inference on mp4 video file (with or without stream=True):

`results = trained_model.predict(source = vid2, imgsz=640)`

> video 1/1 (1/1458) /path/to/video.mp4: 640x640 unplugged 1.00, plugged 0.00, 2.8ms
> ...
> video 1/1 (1457/1458) /path/to/video.mp4: 640x640 unplugged 1.00, plugged 0.00, 2.6ms
> video 1/1 (1458/1458) /path/to/video.mp4: 640x640 unplugged 1.00, plugged 0.00, 2.6ms
> **Speed: 39.6ms preprocess**, 2.9ms inference, 0.0ms postprocess per image at shape (1, 3, 640, 640)

### Deploying solution (at deployment the video will be rtsp stream instead of mp4, but right now I am benchmarking on mp4):

```
from ultralytics import YOLO
import time

def inference(video_address: str):
    # Load the latest trained model
    model = YOLO('C:\\path\\to\\weights\\best_01.pt')

    # INITIALIZE VARIABLES
    start = time.time()
    results = []
    frame_counter = 0

    loop = model.predict(source = video_address, stream=True, verbose = False)

    # Run YOLOv8 inference on the frame
    for r in loop:
        results.append(r.probs.top1)
        frame_counter+=1

        # If inferencing running for 2000 frames or program runs for greater than 4 mins, stop
        if frame_counter>2000 or time.time() - start > 240:
            print(f"Timeout received. Saving results for {video_address}")
            break

    print(f"Request {video_address[:-4]}: Total Time taken: ", time.time() - start)

```

#### Getting results on running a 49 seconds video
> Request 49sec_video: Total Time taken:  129.20500016212463

When setting vid_stride = 3 (effectively making the mp4 run at 10 fps)
> Request 49sec_video_stride3: Total Time taken:  47.64977526664734

When setting vid_stride = 3, verbose= True, stream_buffer=True 
> video 1/1 (486/486) C:\path\to\video.mp4: 640x640 unplugged 1.00, plugged 0.00, 0.0ms
> **Speed: 56.2ms preprocess**, 3.5ms inference, 0.1ms postprocess per image at shape (1, 3, 640, 640)
> Request 49sec_video_stride3_verbose_steambuff: Total Time taken:  44.591055154800415

#### Screenshot of nvidia-smi + task manager when program is running
![image](https://github.com/ultralytics/ultralytics/assets/25133376/6511ecd8-277f-4ffa-bf47-4750f71b0029)

## _Questions:_
1.	**Is there a way to speed up pre-processing bottleneck?** The model is running on CUDA. However (in my opinion), the pre-processing of taking the video, getting its frame, converting it from 2688x1520 to 640x640 and setting it up as input to the model takes place in the CPU and becomes the bottleneck.
2.	**Should I retrain the model with better and/or different arguments?** I thought rectangular sized images/frames is slower to inference hence I went with default 640x640 training and inferencing. Should I try 16x smaller resolution: 672x380 (instead of 2688x1520) for rectangular training and inference.

### Some info that will help:
1.	I cannot change the resolution of raw training videos and rtsp video (when put in use). It will stay at 2688x1520 @ 30fps. I can use vid_stride=3 but the temporal resolution is affected.
2.	I am happy with my training and test/deployment accuracy. The classification is really accurate even with 640 resolution, but preprocessing is the only part I want to work on.
4.	In RTSP, the program automatically starts through a command and runs until stopping condition of 2000 frames is met.
5.	I cannot use TensorRT or OpenVINO models. The model runs of NVIDIA L4 when deployed but the OS is Windows (getting TensorRT is tricky).
6.	I am okay with retraining, changing arguments etc. But the hardware and OS stays fixed. The system is primarily used for just this task, so it is okay allocating all the resources to YOLO.
7.	CPU: Intel® Xeon® D-2775TE, RAM: 64GB @3200MHz, GPU: Nvidia L4 24GB

### TL:DR
Inferencing on YOLOv8s-cls works great, but preprocessing takes 60ms per frame of video which causes issues when trying to run real-time. How can I speed up preprocessing when the video is 2688x1520 at 30fps but model is trained for 640x640.

**Comment by Burhan-Q** [comment]
@Sachin-Wani just a quick couple of points to share. 
1. When checking inference times, it's always a good idea to warm up the GPU/model prior. The first few inference calls _always_ have a higher (overall) processing time if not sufficiently warmed up. I arbitrarily use something like:

```py
dummy = np.random.randint(0, 255, (640, 640, 3), np.uint8)  # update with your image size
_ = [model.predict(dummy) for _ in range(15)]  # warm up for 15 inference calls
```
which for my RTX2060 and `yolov8n` model is sufficient as a warm up, but you might find a different/better values for your model and hardware. 

2. TensorRT 10 is _a lot_ easier to install on Windows. You can now just run a `pip` install (I have Windows 10) and it should work. It won't help the preprocessing speeds necessarily but it'll help with the overall inference speeds.

3. A smaller resolution will help with inference times, but not really with preprocessing times.

4. If I recall correctly, rectangular inference is supported for PyTorch models, but not on exported ones. I wouldn't recommend rectangular training if it's not absolutely necessary. 

5. You could also explore other options for ingesting your video stream, but I can't offer much advice here as I don't have any experience. 

Overall I think when you do your timing checks, make sure to warm up the model every time, otherwise you'll see _very_ high overall process times early on. Secondly, I suggest exploring the use of TensorRT as the recent update _has_ made it easier to install on Windows.

**Comment by Sachin-Wani** [comment]
Hi @Burhan-Q , thanks for sharing your insights. 

1. I usually do follow warming up the model couple of times for feeding in videos which helps me get consistent results (in terms of time taken).
2. Thanks for this tip, I will certainly look into getting TensorRT, although its only going to speed up my inferencing.
3. I was thinking since preprocessing is done through OpenCV (at the backend), is it possible to speed up the bottleneck of getting a frame and converting it into a format usable by the model?

Right now it takes me **56.2ms preprocess**, 3.5ms inference, 0.1ms postprocess per image/frame.

I am not sure if this makes sense, but is there a way to do all the preprocessing on the GPU instead of CPU?

Looking at the preprocessing step in [predict.py ](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/classify/predict.py)

Which essentially does:
```
def preprocess(self, img):
    img = torch.stack([self.transforms(Image.fromarray(cv2.cvtColor(im, cv2.COLOR_BGR2RGB))) for im in img], dim=0)
    img = (img if isinstance(img, torch.Tensor) else torch.from_numpy(img)).to(self.model.device)
    return img.half() if self.model.fp16 else img.float()  # uint8 to fp16/32
```

is this process the source of bottleneck or would it be the [dataloader](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/data/loaders.py)?

For now using vid_stride 3 makes it close to real time in rtsp stream but it'd be great to know if there's a way to speed up the image transforms.

Thanks again for the help and support!

P.S. I see users who have object detection running in real-time, I wonder if they have a better way on handling the data or using resolution that exactly matches the model input.

**Comment by Burhan-Q** [comment]
@Sachin-Wani so I ran a quick test using a public stream I found:

```py
import time
from ultralytics import YOLO


def main():
    src = "http://77.222.181.11:8080/mjpg/video.mjpg"
    # 1/1: http://77.222.181.11:8080/mjpg/video.mjpg... Success ✅ (inf frames of shape 800x500 at 25.00 FPS)
    model = YOLO('yolov8n.pt')
    inference_stream = model.predict(src, stream=True, verbose=False)

    frames = 0
    t0 = time.time()
    times = []
    for result in inference_stream:
        frames += 1
        print(f"{result.speed = }")
        times.append(result.speed)

        if frames > 120 or time.time() - t0 > 120:
            break
    print("overall average process time: ", sum([sum(s.values()) for s in times]) / len(times))

if __name__ == "__main__":
    main()
```

Most of the time for my system was waiting on the stream to provide the next frame. The average process time at the end of a single test was $50 \text{ms}$ and as you can see I didn't do any warm up. The last few of preprocessing times in the terminal were around $2.0$ - $5.0 \text{ms}$.

Looking back at your original post, I noticed the CPU you're using has a very low max turbo clock speed of $3.10 \text{GHz}$ and base clock of $2.00 \text{GHz}$ which compared to my (significantly cheaper) CPU which can boost up to $4.80 \text{GHz}$ and has a base clock speed of $4.10 \text{GHz}$. I'd suspect (not certain) that the bottleneck could be the CPU itself. If you have another system you can test on that has a higher base/boost clock, I'd do a test there to see, because I can only speculate.

**Comment by Sachin-Wani** [comment]
Hi @Burhan-Q , thanks for your suggestion. While I don't have a different system to run my code, just as a comparison, I ran the same program you have (with same address) in my PC. I ended up having this in my results (last 3 lines).

> WARNING ⚠️ Waiting for stream 0
> result.speed = {'preprocess': 0.0, 'inference': 15.627861022949219, 'postprocess': 0.0}
> overall average process time:  18.615119713397064

Just the first time it took 375ms, then it was around 15-16ms. My preprocessing was always 0.0


For the sake of clarity, I also changed the yolov8n to yolov8n-cls (detection to classification and got these):

> WARNING ⚠️ Waiting for stream 0
> result.speed = {'preprocess': 15.516042709350586, 'inference': 0.0, 'postprocess': 0.0}
> overall average process time:  15.364925816374004

In case of classifications, the number was swinging between preprocess and inference which was interesting to note.

I am not sure if its the CPU that's an issue looking at these numbers. Hope this helps getting closer to our solution.
Thanks again!

**Comment by Sachin-Wani** [comment]
Hi @Burhan-Q ,

1.  Just to test on higher resolution videos, I tried your code snippet on a random 1440p video ("https://www.youtube.com/watch?v=fDm8JdCAPeM")

    - In case of classification, I had preprocess time somewhere in 35-55ms range and inference jumping between 0.0-15.6ms, nothing in between.
        
        >overall average process time:  46.97216760127916

    -  In case of using yolov8n.pt (object detection), I had little to no preprocessing time while inference was similar at 15.6ms. _It makes me wonder why so much pre-processing time is needed in case of classification where object detection is super fast?_
        > overall average process time:  16.577196936322075
    

2. I also tried with mp4 files. Same video downloaded at 1080p and ran through your code snippet.

    - In case of classification, there was a significant preprocessing overhead. **CPU usage was at 60%**. (here are last lines of output)
      >       result.speed = {'preprocess': 31.244277954101562, 'inference': 0.0, 'postprocess': 0.0}
      >       result.speed = {'preprocess': 15.62190055847168, 'inference': 0.0, 'postprocess': 0.0}
      >       result.speed = {'preprocess': 16.129732131958008, 'inference': 15.507936477661133, 'postprocess': 0.0}
      >       result.speed = {'preprocess': 31.137704849243164, 'inference': 0.0, 'postprocess': 0.0}
      >       result.speed = {'preprocess': 31.134605407714844, 'inference': 0.0, 'postprocess': 0.0}
      >       result.speed = {'preprocess': 15.62643051147461, 'inference': 0.0, 'postprocess': 0.0}
      >       overall average process time:  28.341738494126112

    - In case of detection, it was super fast again, nearly no preprocess time, and 15.6ms inference. **CPU usage was 0.2%** (here are last lines of output)
      >     result.speed = {'preprocess': 0.0, 'inference': 15.62047004699707, 'postprocess': 0.0}
      >     result.speed = {'preprocess': 0.0, 'inference': 0.0, 'postprocess': 0.0}
      >     result.speed = {'preprocess': 0.0, 'inference': 15.62047004699707, 'postprocess': 0.0}
      >     result.speed = {'preprocess': 0.0, 'inference': 15.623331069946289, 'postprocess': 0.0}
      >     result.speed = {'preprocess': 15.621423721313477, 'inference': 0.0, 'postprocess': 0.0}
      >     result.speed = {'preprocess': 0.0, 'inference': 15.62190055847168, 'postprocess': 0.0}
      >     result.speed = {'preprocess': 0.0, 'inference': 0.0, 'postprocess': 0.0}
      >     overall average process time:  14.970298532720332
 
 
### Can you try using yolov8-cls models and see why that's the case? I thought classification should be faster than object detection, but it looks opposite through YOLO. Maybe this can highlight something that can be improved in the repository internally?

Thanks and happy to provide/test any additional information.

**Comment by Sachin-Wani** [comment]
Running the code with verbose, I noticed that for my videos the detection (yolov8n) runs at 384x640. 
> Speed: 2.9ms preprocess, 11.0ms inference, 1.6ms postprocess per image at shape (1, 3, 384, 640)

Using the same video on yolov8n-cls the classification runs at 224x224 with preprocessing being the overhead causing cpu usage.
> Speed: 45.7ms preprocess, 4.5ms inference, 0.1ms postprocess per image at shape (1, 3, 224, 224)

Using the same video on my trained classification model which runs at 640x640
> Speed: 70.1ms preprocess, 4.3ms inference, 0.1ms postprocess per image at shape (1, 3, 640, 640)

### Is there a way to eliminate this bottleneck in classifications (which I think is caused by OpenCV transforms)?

**Comment by glenn-jocher** [comment]
Hi @Sachin-Wani,

Thank you for the detailed analysis and the additional tests you conducted. Your observations are indeed insightful and highlight a key difference in preprocessing times between classification and detection models.

### Addressing the Preprocessing Bottleneck

1. **GPU-Based Preprocessing**:
   As you correctly pointed out, the preprocessing step, especially the image resizing and transformations, is currently handled on the CPU. One potential way to speed this up is to leverage GPU-based preprocessing. Libraries like NVIDIA DALI (Data Loading Library) can help offload these tasks to the GPU, significantly reducing the preprocessing time. Here's a basic example of how you might integrate DALI into your pipeline:

   ```python
   from nvidia.dali.pipeline import Pipeline
   import nvidia.dali.fn as fn
   import nvidia.dali.types as types
   from nvidia.dali.plugin.pytorch import DALIGenericIterator
   import torch

   class ExternalInputIterator:
       def __init__(self, batch_size, image_paths):
           self.batch_size = batch_size
           self.image_paths = image_paths
           self.n = len(image_paths)

       def __iter__(self):
           self.i = 0
           return self

       def __next__(self):
           batch = []
           labels = []
           for _ in range(self.batch_size):
               img = cv2.imread(self.image_paths[self.i])
               batch.append(img)
               labels.append(0)  # Dummy label
               self.i = (self.i + 1) % self.n
           return (batch, labels)

   class SimplePipeline(Pipeline):
       def __init__(self, batch_size, num_threads, device_id, external_data):
           super(SimplePipeline, self).__init__(batch_size, num_threads, device_id, seed=12)
           self.input = fn.external_source(source=external_data, num_outputs=2)
           self.decode = fn.decoders.image(device="mixed", output_type=types.RGB)
           self.resize = fn.resize(device="gpu", resize_x=640, resize_y=640)
           self.cmnp = fn.crop_mirror_normalize(device="gpu",
                                                dtype=types.FLOAT,
                                                output_layout="CHW",
                                                crop=(640, 640),
                                                mean=[0.485 * 255, 0.456 * 255, 0.406 * 255],
                                                std=[0.229 * 255, 0.224 * 255, 0.225 * 255])

       def define_graph(self):
           self.jpegs, self.labels = self.input()
           images = self.decode(self.jpegs)
           images = self.resize(images)
           output = self.cmnp(images)
           return output, self.labels

   batch_size = 32
   image_paths = ["path/to/image1.jpg", "path/to/image2.jpg", ...]
   eii = ExternalInputIterator(batch_size, image_paths)
   pipe = SimplePipeline(batch_size=batch_size, num_threads=2, device_id=0, external_data=eii)
   pipe.build()
   dali_iter = DALIGenericIterator(pipe, ["data", "label"], size=len(image_paths))

   for data in dali_iter:
       images = data[0]["data"]
       labels = data[0]["label"]
       # Use images and labels for your model
   ```

2. **Optimizing OpenCV Operations**:
   If integrating DALI is not feasible, you might want to look into optimizing your OpenCV operations. For instance, using multi-threading or parallel processing to handle the frame extraction and resizing can help distribute the load across multiple CPU cores.

3. **Batch Processing**:
   Another approach is to process frames in batches rather than individually. This can help reduce the overhead associated with repeatedly calling the preprocessing functions.

4. **Model Input Size**:
   While you mentioned that you are happy with the accuracy at 640x640, experimenting with slightly lower resolutions (e.g., 512x512) might provide a good balance between preprocessing time and model performance.

### Example Code for Batch Processing with OpenCV

Here's an example of how you might implement batch processing with OpenCV:

```python
import cv2
import numpy as np
from ultralytics import YOLO

def preprocess_batch(frames, target_size=(640, 640)):
    processed_frames = []
    for frame in frames:
        resized_frame = cv2.resize(frame, target_size)
        processed_frames.append(resized_frame)
    return np.array(processed_frames)

def main():
    src = "path/to/video.mp4"
    model = YOLO('yolov8n-cls.pt')
    cap = cv2.VideoCapture(src)
    batch_size = 32
    frames = []

    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            break
        frames.append(frame)
        if len(frames) == batch_size:
            processed_frames = preprocess_batch(frames)
            results = model.predict(processed_frames)
            frames = []

    cap.release()

if __name__ == "__main__":
    main()
```

### Conclusion

By leveraging GPU-based preprocessing, optimizing OpenCV operations, and processing frames in batches, you can significantly reduce the preprocessing bottleneck and achieve near real-time performance for your classification tasks. Feel free to experiment with these approaches and let us know how it goes!

If you have any further questions or need additional assistance, please don't hesitate to ask. We're here to help! 🚀

**Comment by Sachin-Wani** [comment]
Hi @glenn-jocher , Thank you for this information!

My final goal is to use the solution for RTSP streams. So, I am not sure if using batch processing will help as collecting frames is still going to take some time. But DALI could be helpful reduce CPU strain and speed things up. I am not sure how to use it for RTSP frames and currently my pipeline relies much on YOLO's repository for much of background work:
```
    loop = model.predict(source = video_address, stream=True, verbose = False, vid_stride = 3, stream_buffer = True)

    # Run YOLOv8 inference on the frame
    for r in loop:
        results.append(r.probs.top1)
        frame_counter+=1

        # If inferencing running for 2000 frames or program runs for greater than 4 mins, stop
        if frame_counter>2000 or time.time() - start > 240:
            print(f"Timeout received. Saving results for {video_address}")
            break
```

My solution's goal is to count the number of frames that are 0/1 (plug/unplug), so I would ideally not want to skip any frames (hence added the stream_buffer), and would want the stride to be 1. However given the current scenario of preprocessing bottleneck, I am using vid_stride = 3 which makes the video equivalent to 10fps.

#### I am just wondering, why doesn't the object detection go through the same resizing and transformation phase through cv2 regardless of the media input resolution?

As for changing the model input size, that will help reduce some amount of total time taken, resizing & normalizing the image would still take time given the high original resolution.

Thanks again.

**Comment by Burhan-Q** [comment]
@Sachin-Wani I reran the snippet (with the same RTSP video link) using `yolov8n-cls.pt` and the overall average process time for me was $6.857 \text{ ms}$ and typical inference looked like

```
{'preprocess': 1.0004043579101562, 'inference': 3.000497817993164, 'postprocess': 0.0}
{'preprocess': 1.0035037994384766, 'inference': 5.995988845825195, 'postprocess': 0.0}
{'preprocess': 1.0020732879638672, 'inference': 8.997201919555664, 'postprocess': 0.0}
```
the worst preprocess time I saw was $3.998 \text{ ms}$

**Comment by Burhan-Q** [comment]
What version of OpenCV are you on btw? The PC I just test with is running `opencv-python==4.8.1.78`

**Comment by Burhan-Q** [comment]
@Sachin-Wani using that video link, here are the results from my machine (at the model native 224 x 224 inference size). 

```
overall average process time:  4.489 ms
average preprocess time: 1.018 ms
average inference time: 3.455 ms
average postprocess time: 0.017 ms
```

With respect to your question about the preprocessing for classification versus object detection models, I don't have an _exact_ answer, but I tried changing the existing preprocessing step for the `classify` task with something more similar to the `detect` class and the times were not much different than what I have here. I didn't monitor my CPU usage closely, but I do have a running plot and I don't think that the usage would have been notable compared to all the other activity running on my PC.

**Comment by Sachin-Wani** [comment]
@Burhan-Q , that is really interesting!
In my case as mentioned above using the youtube link (2560x1440 x30fps) through yolov8n-cls classification at 224x224 I had preprocess time somewhere in 35-55ms range and inference jumping between 0.0-15.6ms, nothing in between.
overall average process time: 46.97216760127916

Looking at your numbers make me wonder if I am using something incorrectly. My CPU usage shoots up when using higher resolution video streams (compared to the public rtsp you shared). With your numbers, that is more than capable of running real-time. For me, lower resolution input streams/videos work much faster as there's not as much preprocessing needed. However with a larger resolution if you are getting such good numbers in classification preprocessing, I might have to look into what's causing my system to take up such time. 

Thanks for sharing this.

**Comment by Burhan-Q** [comment]
@Sachin-Wani just a silly question...have you checked to make sure you have `torch` with CUDA and have the inference running on GPU? I've assumed that you have up to now, but sometimes it's worth asking the "obvious" question. Could be worth running a basic example with images as well, you could try running `val` with `data="imagenet10"`

**Comment by Sachin-Wani** [comment]
Hi @Burhan-Q ,
I am fairly positive that the model is running on GPU. I have CUDA in torch:
![image](https://github.com/ultralytics/ultralytics/assets/25133376/71631737-2989-4f4f-8d3e-6ef455afaff9)

While I am running the model, I see python.exe in my nvidia-smi task and also have supporting nvcc (12.1). While training I had something like this indicating its training on the GPU.
```
  Epoch    GPU_mem       loss  Instances       Size
  50/50       4.5G    0.01264          7        640: 100%|██████████| 1233/1233 [07:10<00:00,  2.87it/s]
           classes   top1_acc   top5_acc: 100%|██████████| 30/30 [00:12<00:00,  2.49it/s]
               all      0.971          1
```

 I am not sure how to check for inferencing but this is what I got running val from CLI. I also used nvidia-smi at the same time to see if GPU is used:
 
![image](https://github.com/ultralytics/ultralytics/assets/25133376/bca32400-0b61-4b96-84f6-5ef38856ed69)


Hope the above information helps. 

### I am trying to get access to a desktop PC with Intel i7 and RTX GPU to see if CPU is causing the bottleneck. I will use the latest version of everything supported on that system and will follow back with my observations.

**Comment by Burhan-Q** [comment]
@Sachin-Wani sounds good, and yeah I assumed you had checked all that, but just wanted to be 100% certain 😆. Your screenshot shows that you're using `CUDA:0` when running validation, so that's good. Certainly curious to hear how the testing goes with another machine, and maybe consider using the Docker container (if possible) so you can be certain that the environment is configured the same.

**Comment by Sachin-Wani** [comment]
Hi @Burhan-Q ,
Just to follow up on getting a new generation PC to test my classification, I am getting a 13th gen i9 and rtx 40 series gpu for a week. Just to make sure I install everything correctly, here are the steps I follow (if you have any different method, please let me know):

1. Install NVIDIA Drivers - latest from their website. Confirm with nvidia-smi command.
2. Install miniconda and create a new environment and activate with 
3.  `conda create -n yolo python=3.11`
4. Install cuda drivers (I prefer 12.1 as pytorch supports it) with
5. `conda install nvidia/label/cuda-12.1.0::cuda`
6. Check with nvcc --version in the conda environment
7. Install pytorch with cuda
8. `pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121`
9. Verify using torch.cuda.is_available() in python
10. Finally install ultralytics and any additional libraries with pip install

Once I am done with the initial setup, I will run some benchmarks and compare it with my results. I will post my findings here.
Thanks for the help as always!

**Comment by Burhan-Q** [comment]
@Sachin-Wani I'm jealous of the PC! 😆 

I'm not a big `conda` user personally, but that all looks correct to me.
